src/Confluent.Kafka/AdminClient.cs

```diff
@@ -501,6 +501,17 @@ public Metadata GetMetadata(string topic, TimeSpan timeout)
         public int AddBrokers(string brokers)
             => kafkaHandle.AddBrokers(brokers);
 
+        /// <summary>
+        ///     Refer to <see cref="Confluent.Kafka.IClient.OauthBearerSetToken(string, long, string, IDictionary{string, string})" />
+        /// </summary>
+        public void OauthBearerSetToken(string tokenValue, long lifetimeMs, string principalName, IDictionary<string, string> extensions)
```

Comment:
I don't like having these methods exposed at the client class level because they aren't generally applicable - putting them here messys up the name space for all users of the clients, who mostly won't care about them. I think it would be better to pass in two additional delegate arguments into the refresh handler instead.

Reply:
The idea is that tokens can be set/failed at finer granularity than the refresh frequency, so these need to be exposed as standalone methods.

Reply:
yes, but better if these methods are passed into the handler to limit their scope.

Reply:
oh, that's annoying

Reply:
is the requirement simply that a higher rate of refresh than the default assumed by librdkafka may be desired (which is 80% of lifetime)? or might the user want to do this on an ad-hoc basis? If it's the former, this could be achieved via configuration (and i like it better). if the latter, what's the scenario?

Reply:
Look at the original librdkafka PR for motivation.

Reply:
that has 321 comments.. i think we can have a chat.

Reply:
@thtp - i discussed with @edenhill and we agreed my suggestion is the more conservative. if it turns out the client level methods are required, we can retrofit easily. do you specifically require this yourself?

so on the builder classes, i'm thinking something like:

```
public ConsumerBuilder<TKey, TValue> SetOauthBearerTokenRefreshHandler(Action<IConsumer<TKey, TValue>, SetTokenDelegate, SetTokenErrorDelegate, string> oauthBearerTokenRefreshHandler)
```

(if we ever add oauth methods to the client we could overload this with a `SetOauthBearerTokenRefreshHandler` as you have it as well).

Reply:
My main thoughts was that someone might have scenarios with reactive token revocation, or might need to update claims, or something like this. I mean, there might be reasons to update or revoke token besides its expiration.
Also, IClient wraps RdKafka HandleImpl, and oauthbearer_set_token is declared in than HandleImpl class. So it just feels natural to place it there (just mimicking original api).
The last one - just clean and easy to use method signature. Donno, I just like how it looks. Less code written, less delegates declaration, less ctrl+clicking around the solution.

Well, I bet you've discussed all that already...
So, delegates it be? Final decision? 146% sure? ) @mhowlett ?

p.s: I hope my attempts to express my thoughts in English are at least understandable )

Reply:
i will think about this more, give me a day or two.

Reply:
some thoughts:

> The last one - just clean and easy to use method signature. Donno, I just like how it looks. Less code written, less delegates declaration, less ctrl+clicking around the solution.

What you suggest is more straightforward / simpler for people to get, and that is absolutely a big advantage. What I proposed does add UX complexity for the sake of a cleaner separation of concerns. This is not always/necessarily a good tradeoff.

> My main thoughts was that someone might have scenarios with reactive token revocation, or might need to update claims, or something like that. I mean, there might be reasons to update or revoke token besides its expiration.

My assumption was this isn't a super important capability because AFAICT, the java implementation doesn't allow for this (am i missing something here?). that's not to say it isn't useful though. The scheme i proposed does actually allow for this, since the delegate instance lifetime can extend beyond that of the handler. not very good UX (but if not commonly needed, not something to optimise for).

> Also, IClient wraps RdKafka HandleImpl, and oauthbearer_set_token is declared in than HandleImpl class. So it just feels natural to place it there (just mimicking original api).

It's not an is-a relationship.

Reply:
Wondered if extension methods can help. Maybe, but it's not very clean.

First you'd have:

```
namespace Confluent.Kafka.Oauthbearer
{
    public static class OauthBearerProducerExtensions
    {
        public static void OauthBearerSetToken<TKey, TValue>(this IProducer<TKey, TValue> producer, string tokenValue, long lifetimeMs, string principalName, IDictionary<string, string> extensions)
        {
            // ...
        }

        public static void OauthBearerSetTokenFailure<TKey, TValue>(this IProducer<TKey, TValue> producer, string error)
        {
            // ...
        }
    }
```

etc.

the client instances have `Handle`, which have a `SafeKafkaHandle`. However you can't get at it in the extension method without using reflection because it's internal. You could use reflection, or define a helper class as follows (which can access internal properties) and use that instead. I think it'd probably go reflection, but it feels a bit horrible.

```
    public static class OauthbearerUtils
    {
        public static void OauthBearerSetToken<TKey, TValue>(Handle handle, string tokenValue, long lifetimeMs, string principalName, IDictionary<string, string> extensions)
        {
            // ...
        }

        public static void OauthBearerSetTokenFailure(Handle handle, string error)
        {
            // ...
        }
    }
}
```

Then we have the problem of how to specify the refresh handler on the builder classes. Again, we have the problem of `internal` properties. i think you could get around that somewhat by having an extension method on the `ProducerBuilder` etc. for adding the refresh handler, which returns `OauthBearerProducerBuilder`, but this a bit of a hack and isn't great because the same trick won't work if we want extensibility for something else. Instead, I think we'd need a general public method on the builder classes: `SetHandler(string name, object handler)`, but this is not very good / overly complex as well.

Reply:
current thinking:

1. extension methods don't help.
2. i don't like either approach.
3. i do still currently prefer what i suggested, but i'm not very confident about it.

please bear with me a bit longer ...

Reply:
another option is for the handler to return an object which represents either a valid token or an error. 

```
public ConsumerBuilder<TKey, TValue> SetOAuthBearerTokenRefreshHandler(Func<IConsumer<TKey, TValue>, string, OAuthBearerRefreshResult> oAuthBearerTokenRefreshHandler)

OAuthBearerToken : OAuthBearerRefreshResult
OAuthBearerError: : OAuthBearerRefreshResult
```

this doesn't allow for reactive token renewal / cancellation, but it's a step down in UX complexity. It's also arguably an improvement since it forces the behavior in the handler that is required - a single action to set the token or set the token error.

also, there is nothing preventing us from expanding the capability in the future if required - this doesn't cut off any avenues for extensibility.

this is currently my favourite option.

Reply:
I think we'll want to be cautious in having one of the librdkafka-based clients do things differently from the others, and only do so on solid grounds.
In this particular case I don't see any reason, other than aesthetics, from deviating from librdkafka, and in fact it reduces functionality of a feature that by nature requires a lot of flexibility.

Reply:
Actually, extensions might be a good solution. I mean, extensions might be provided by this library. So, internal methods will not be a problem, we can leave IClient without any mods, and still have an ability to set token using IClient instance.

Reply:
@thtp i was excited by extension methods too. unfortunately it turns out they don't have access to internal members :-(. we could use reflection in both cases to get access to internal members. this is a bit horrible, but the user is not aware it's happening.

@edenhill - take your point. I think it too, i'm pretty torn on this.

(currently letting this sit for a bit - some chance what is right will present itself with certainty).

Reply:
ok, i'm coming around to thinking as is wins (80% sure). let's let it sit another day though.

Reply:
@mhowlett check out extensions implementation. Looks like perfect compromise ) (will revert if it's bulls..t)

Reply:
interesting. i didn't think that would compile (based on what i was reading), but it does. we'll need to do the same trick on the builder classes, because it doesn't make sense to be able to set the refresh handler if the methods don't exist on the clients to implement them...

it is seeming good, yes. i think it's worth implementing those builder class changes (i look forward to taking it for a spin and seeing how it feels).

note: I do think it's worth dwelling on these things for a bit, because once you decide on something you're stuck with it and sometimes seemingly insignificant structural decisions can end up ballooning into much bigger things in the future.

Reply:
Well, modifying the builders might be a little bit much more difficult... 
I mean, builders does not have a reference to a Handle, so we can't call SetWhateverHandler in a builder extension. We have to store delegate somewhere before it can be called on a Handler instance. Extension properties does not make it to the C#8, so the best we can do is storing delegates as a Dictionary<object, object> in the static field of an extension class. And we still will need to call a method to set those handlers in the ctor of the client.
Another way is to make it an extension on IClient. So the user will have to create an instance of a client with builder, and then on that instance call SetWhateverHandler().
Both options look weird.
To make it less weird we will have to change builders dramatically. It might become just a chane of Action<TClient> delegates that will initialize  an instance of a client or something like that, whatever... Anyway, it have to be changed if you want to remove any knowledge about all the handlers from a builders.

Reply:
the builders could have an internal OAuthBearerTokenRefreshHandler property, but there's no way to set it without the extension method is what i was thinking